File: Python/data_structures/heaps/max_heap.py
class MaxHeap(object):
    """

    Implementation of max heap with list data structure.
    Operations:
    insert(value)
    delete(value)
    sift_down(value)

    Linked list complexity:
    [method]: [worst case]
    largest: O(1)
    insert: O(log n)
    delete: O(log n)

    """
    def __init__(self, heap=list()):
        self.heap = heap

    def largest(self):
        """

        Return the largest value in the heap.

        """
        return self.heap[0]

    def insert(self, value):
        """

        Insert value into the maxheap structure and return the heap.

        """
        if len(self.heap) < 1:
            self.heap.append(value)
            return
        self.heap.append(value)
        # exchange top element with new value
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        #maintain max heap
        self.heap = max_heapify(self.heap)
        return

    def remove(self):
        """

        Remove the root(item with the largest value) from heap.
        Return the removed value.

        """
        # move last element to first position and vice versa
        self.heap[0], self.heap[-1] = self.heap[-1], self.heap[0]
        value = self.heap.pop()
        # maintain max heap
        self.heap = max_heapify(self.heap)
        return value

    def left_child_index(self, index):
        return 2*index + 1

    def right_child_index(self, index):
        return 2*index + 2

    def parent_index(self, index):
        return int(index/2)


def sift_down(heap, index):
    """

    sift_down assumes that binary trees rooted at left and right of index
    element are max heaps. but heap[index] might be smaller than it's children,
    thus violating the max heap property. sift_down lets the value float down
    in the max heap so that the subtree rooted at index obeys the max heap
    property.

    """
    left = 2*index + 1
    right = 2*index + 2
    largest = index
    if left < len(heap) and heap[left] > heap[largest]:
        largest = left
    if right < len(heap) and heap[right] > heap[largest]:
        # right child largest
        largest = right
    if largest != index:
        heap[index], heap[largest] = heap[largest], heap[index]
        return sift_down(heap, largest)
    return heap


def max_heapify(a_list):
    """

    Build a max heap from a given list.
    O(n lgn)

    """
    for i in range(len(a_list)/2, -1, -1):
        sift_down(a_list, i)
    return a_list


if __name__ == "__main__":
    test_list = [7, 19, 36, 17, 3, 25, 1, 2, 100]
    print max_heapify(test_list)

    max_heap = MaxHeap()
    max_heap.insert(7)
    max_heap.insert(19)
    max_heap.insert(36)
    max_heap.insert(17)
    max_heap.insert(3)
    max_heap.insert(25)
    max_heap.insert(1)
    max_heap.insert(2)
    max_heap.insert(100)
    print max_heap.heap
    max_heap.remove()
    print max_heap.heap
    max_heap.remove()
    print max_heap.heap

File: PHP/data_structures/heaps/max_heap.php
<?php

/**
 * Implementing a min heap with array.
 */

class MaxHeap {
    private $heap;

    public function __construct($heap=array()) {
        $this->heap = $heap;
    }

    /**
     * Insert element into the heap.
     * Time: O(lg n). where n is size of the heap
     */
    public function insert($data) {
        $this->heap[] = $data;
        $i = sizeof($this->heap) - 1;
        $parentIndex = $this->getParentIndex($i);

        while ($parentIndex >= 0 && $i > 0) {
            if ($this->heap[$parentIndex] < $data) {
                $this->heap[$i] = $this->heap[$parentIndex];
            } else {
                break;
            }
            $i = $parentIndex;
            $parentIndex = $this->getParentIndex($i);
        }
        $this->heap[$i] = $data;
    }

    /**
     * Delete the largest element from the heap.
     * Time: O(lg n). where n is size of the heap
     */
    public function delete() {
        // check empty heap
        if (empty($this->heap)) {
            return false;
        }
        // swap first and last index
        $this->swap($this->heap[0], $this->heap[sizeof($this->heap)-1]);

        // delete last index element
        $largest = array_pop($this->heap);
        // siftdown the new first index
        if (!empty($this->heap)) {
            $this->siftDown(0);
        }
        return $largest;
    }

    private function swap(&$value1, &$value2) {
        $temp = $value1;
        $value1 = $value2;
        $value2 = $temp;
    }

    /**
     * Make the heap a max heap.
     * Time: O(n)
     */
    public function heapify() {
        if (empty($this->heap)) {
            return;
        }
        // get parent node of the last node
        $parentIndex = $this->getParentIndex(sizeof($this->heap)-1);
        // from the parent node until root node, repeat sift down
        while ($parentIndex >= 0) {
            $this->siftDown($parentIndex);
            $parentIndex--;
        }
    }

    /**
     * Sift down value at given index repeatedly so the max heap structure is maintained.
     * Time: O(lg n). where n is size of the heap
     */
    private function siftDown($siftIndex) {
        if ($siftIndex >= sizeof($this->heap)) {
            return;
        }
        // set temp for sift index value
        $temp = $this->heap[$siftIndex];

        $i = $siftIndex;
        $childIndex = $this->getLeftChildIndex($i);

        // while loop on there is a left child
        while ($childIndex < sizeof($this->heap)) {
            // right child bigger than left child
            if ($childIndex+1 < sizeof($this->heap) &&
                $this->heap[$childIndex+1] > $this->heap[$childIndex]) {
                $childIndex++;
            }
            if ($temp < $this->heap[$childIndex]) {
                $this->heap[$i] = $this->heap[$childIndex];
            } else {
                break;
            }
            $i = $childIndex;
            $childIndex = $this->getLeftChildIndex($i);
        }
        // insert temp into the correct spot
        $this->heap[$i] = $temp;
    }

    public function getHeap() {
        return $this->heap;
    }

    private function getParentIndex($index) {
        if ($index == 0) {
            return NULL;
        }
        return floor(($index-1) / 2);
    }

    private function getLeftChildIndex($index) {
        return $index*2 + 1;
    }

    private function getRightChildIndex($index) {
        return $index*2 + 2;
    }
}

$presetArray = [104,71,24,66,27,23,8,5,32,25,18,22];
$nonMaxHeapArray = [3,6,10,18,8,7,25,400];
$maxHeap = new MaxHeap($nonMaxHeapArray);

// print_r($maxHeap->getHeap());
// echo $maxHeap->delete();
// print_r($maxHeap->getHeap());
// echo $maxHeap->delete();
// print_r($maxHeap->getHeap());

// $maxHeap->insert(50);
// print_r($maxHeap->getHeap());
// $maxHeap->insert(91);
// print_r($maxHeap->getHeap());
// $maxHeap->insert(150);
// print_r($maxHeap->getHeap());

// $maxHeap->heapify();
// print_r($maxHeap->getHeap());
